lib/common/schemas.c

```diff
@@ -1074,41 +1074,48 @@ update_validation(xmlNode **xml_blob, int *best, int max, gboolean transform,
 
         if (rc == pcmk_ok && transform) {
             xmlNode *upgrade = NULL;
+            struct schema_s *next_schema = NULL;
             next = schema->after_transform;
 
             if (next <= lpc) {
                 /* There is no next version, or next would regress */
                 crm_trace("Stopping at %s", schema->name);
                 break;
+            }
 
-            } else if (max > 0 && (lpc == max || next > max)) {
+            if (max > 0 && (lpc == max || next > max)) {
                 crm_trace("Upgrade limit reached at %s (lpc=%d, next=%d, max=%d)",
                           schema->name, lpc, next, max);
                 break;
+            }
+
+            if (next >= 0) {
+                next_schema = &known_schemas[next];
+            }
 
-            } else if (schema->transform == NULL
+            if (schema->transform == NULL
                        /* possibly avoid transforming when readily valid
                           (in general more restricted when crossing the major
                           version boundary, as X.0 "transitional" version is
                           expected to be more strict than it's successors that
                           may re-allow constructs from previous major line) */
-                       || validate_with_silent(xml, next >= 0 ? &known_schemas[next] : NULL)) {
+                       || validate_with_silent(xml, next_schema)) {
                 crm_debug("%s-style configuration is also valid for %s",
-                           schema->name, known_schemas[next].name);
+                           schema->name, next_schema->name);
```

Comment:
next_schema?  Hm, no I don't think so.  The combination of setting it to `lpc+1` earlier in this block and then checking that it's less than max/max_stable_schemas should mean that it's always valid, unless the known_schemas list gets screwed up somehow.  I can add a check and a log message just in case, though.

This has me wondering if setting next to -1 in various places is still necessary, or if I've screwed something up.

Reply:
Okay, the -1 stuff is definitely important and still has a use.  I added a `CRM_ASSERT(next_schema != NULL)` that never gets triggered by our tests, so I think we're okay to leave that in.